### PR TITLE
Remove unused variables

### PR DESCRIPTION
**Pull request by basoundr**
This analyzer is only for VB because in VB, if a local variable is written into then the warning disappears. But we need the warning to go away only when there is a read reference on the local. C# compiler already has the desired behavior to report unused variables properly.

@dotnet/roslyn-analysis Could someone please review this?

Also, the bug I mentioned in the standup that the locals in InvocationExpression where not detected is not correct. I had a compiler error in the testcase which prevented the expression from being bound. I used `Console.WriteLine` without importing `System`

### CODE REVIEW

**Comment by mavasani** [comment]
> Also, the bug I mentioned in the standup that the locals in InvocationExpression where not detected is not correct. I had a compiler error in the testcase which prevented the expression from being bound. I used  Console.WriteLine  without importing  System 

This still seems like a bug. Even though the invocation doesn't bound to a non-error type method symbol, the argument to it is still a valid local reference. Are you suggesting that the compiler is not generating a BoundLocal for the argument?

**Comment by basoundr** [comment]
@mavasani It is true that the argument for the method will still be bound to the Local but the bound expression will be inside an expression which is unbound because of resolution error. At this point, I would assume that we are not getting the bound argument expression(to the local) to be analyzed because it is wrapped by an expression with error. If you think that this behavior is wrong and we would like to analyze all the bound expressions irrespective of its context then I can open an issue.

**Comment by mavasani** [comment]
@basoundr I would still expect a ILocalReferenceExpression for such an local. @JohnHamby ?

**Comment by basoundr** [comment]
@dotnet-bot retest Windows Debug please
// Previous failure: http://dotnet-ci.cloudapp.net/job/dotnet_roslyn-analyzers/job/master/job/windows_debug_prtest/33/
// Retest reason: I/O issue

**Comment by basoundr** [comment]
@dotnet-bot retest Windows Release please
// Previous failure: http://dotnet-ci.cloudapp.net/job/dotnet_roslyn-analyzers/job/master/job/windows_release_prtest/33/
// Retest reason: I/O issue

**Comment by basoundr** [comment]
@dotnet-bot retest default please
// Previous failure: http://dotnet-ci.cloudapp.net/job/dotnet_roslyn-analyzers_prtest/746/
// Retest reason: I/O issue

**Comment by basoundr** [comment]
@dotnet-bot retest default please
// Previous failure: http://dotnet-ci.cloudapp.net/job/dotnet_roslyn-analyzers_prtest/746/
// Retest reason: I/O issues

File: src/Microsoft.Maintainability.Analyzers/Core/RemoveUnusedLocals.cs [hunk 1]
```diff
@@ -23,9 +23,9 @@ public abstract class RemoveUnusedLocalsAnalyzer : DiagnosticAnalyzer
                                                                              s_localizableMessage,
```
Comment (srivatsn) [inline src/Microsoft.Maintainability.Analyzers/Core/RemoveUnusedLocals.cs]: Do we need the base type at all? Can we just the VB type?

Comment (basoundr) [inline src/Microsoft.Maintainability.Analyzers/Core/RemoveUnusedLocals.cs]: That is true. We don't need the base type. I will remove it

**Comment by basoundr** [comment]
@dotnet-bot retest default please
// Previous failure: http://dotnet-ci.cloudapp.net/job/dotnet_roslyn-analyzers_prtest/746/
// Retest reason: Unknown failure

**Comment by JohnHamby** [comment]
Exactly what appears in bound trees for programs with errors is presently subject to the whims of the compilers. For Update 3 (or beyond) we hope to bring the compilers to a state where behavior in the presence of errors is at least systematic and predictable.

**Comment by basoundr** [comment]
@dotnet-bot retest default please
// Previous failure: http://dotnet-ci.cloudapp.net/job/dotnet_roslyn-analyzers_prtest/746/
// Retest reason: CSharp dll was used by another process

**Comment by basoundr** [comment]
@dotnet-bot retest Windows Release please
// Previous failure: http://dotnet-ci.cloudapp.net/job/dotnet_roslyn-analyzers/job/master/job/windows_release_prtest/33/
// Retest reason: I/O issue

**Comment by basoundr** [comment]
@dotnet/roslyn-analysis ping? Finally all the jenkins issues are eliminated and have good results

File: src/Microsoft.Maintainability.Analyzers/VisualBasic/BasicRemoveUnusedLocals.vb [hunk 2]
```diff
@@ -11,7 +14,72 @@ Namespace Microsoft.Maintainability.Analyzers
...
+                        Next
+                    End Sub, OperationKind.VariableDeclarationStatement)
+
+                    operationBlockContext.RegisterOperationAction(
+                    Sub(operationContext)
+                        Dim localReferenceExpression As ILocalReferenceExpression = DirectCast(operationContext.Operation, ILocalReferenceExpression)
+                        Dim syntax = localReferenceExpression.Syntax
+
+                        ' The writeonly references should be ignored
+                        If syntax.Parent.IsKind(SyntaxKind.SimpleAssignmentStatement) AndAlso
```
Comment (mavasani) [inline src/Microsoft.Maintainability.Analyzers/VisualBasic/BasicRemoveUnusedLocals.vb]: @JohnHamby We should consider exposing a property on all reference expressions indicating whether it's a read or write. I presume the bound nodes already have this information.

File: src/Microsoft.Maintainability.Analyzers/UnitTests/RemoveUnusedLocalsTests.cs [hunk 3]
```diff
@@ -14,7 +19,54 @@ protected override DiagnosticAnalyzer GetBasicDiagnosticAnalyzer()
 
         protected override DiagnosticAnalyzer GetCSharpDiagnosticAnalyzer()
         {
-            return new CSharpRemoveUnusedLocalsAnalyzer();
+            throw new NotImplementedException();
+        }
+
+        [Fact]
```
Comment (mavasani) [inline src/Microsoft.Maintainability.Analyzers/UnitTests/RemoveUnusedLocalsTests.cs]: Can you add some more tests: local accessed only in lambda, local passed by ref or out?

**Comment by mavasani** [comment]
:+1:

File: src/Microsoft.Maintainability.Analyzers/UnitTests/RemoveUnusedLocalsTests.cs (hunk 3 above)
Comment (basoundr) [inline src/Microsoft.Maintainability.Analyzers/UnitTests/RemoveUnusedLocalsTests.cs]: Done

**Comment by JohnHamby** [comment]
The bound trees do not currently have enough information to determine whether a reference is used to fetch or to store without knowing the context of the reference. The VB bound trees have a flag that one might believe would carry this information, but it actually means something else, and the C# bound trees have nothing.

**Pull request merged**